File: May-LeetCoding-Challenge/24-Construct-Binary-Search-Tree-From-Preorder-Traversal/Construct-Binary-Search-Tree-From-Preorder-Traversal.kt
class ConstructBinarySearchTreefromPreorderTraversalKotlin1008 {
    fun bstFromPreorder(preorder: IntArray): TreeNode? {
        return construct(preorder, 0, preorder.size - 1)
    }

    private fun construct(preorder: IntArray, start: Int, end: Int): TreeNode? {
        if (start > end || start == preorder.size) {
            return null
        }
        val current = TreeNode(preorder[start])
        var rightIndex = start + 1
        while (rightIndex <= end && current.`val` > preorder[rightIndex]) {
            ++rightIndex
        }
        current.left = construct(preorder, start + 1, rightIndex - 1)
        current.right = construct(preorder, rightIndex, end)
        return current
    }
    /*
    fun bstFromPreorder(preorder: IntArray): TreeNode? {
        if (preorder.isEmpty()) {
            return null
        }
        val root =
            TreeNode(
                preorder[0]
            )
        var index = 1
        while (index < preorder.size && preorder[0] > preorder[index]) {
            ++index
        }
        root.left = bstFromPreorder(preorder.sliceArray(IntRange(1, index - 1)))
        root.right = bstFromPreorder(preorder.sliceArray(IntRange(index, preorder.size - 1)))
        return root
    }
     */

    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }
}

fun main() {
    val array = intArrayOf(8, 5, 1, 7, 10, 12)
    array.sliceArray(IntRange(1, 3)).forEach(::print)
    println()
    array.sliceArray(IntRange(4, 5)).forEach(::print)
}

File: May-LeetCoding-Challenge/24-Construct-Binary-Search-Tree-From-Preorder-Traversal/Construct-Binary-Search-Tree-From-Preorder-Traversal.scala
class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
  var value: Int = _value
  var left: TreeNode = _left
  var right: TreeNode = _right
}

//DFS solution
object Solution {
  def bstFromPreorder(preorder: Array[Int]): TreeNode = {
    var parentList = scala.collection.mutable.Seq.empty[TreeNode]
    val root = new TreeNode(preorder(0))

    def bstFromPreorderWithIndex(index: Int, parent: TreeNode): Unit = {
      if (index != preorder.length) {
        if (parent.value > preorder(index)) {
          parent.left = new TreeNode(preorder(index))
          parentList = parentList.+:(parent)
          bstFromPreorderWithIndex(index + 1, parent.left)
        } else {
          if (parentList.isEmpty || preorder(index) < parentList.head.value) {
            parent.right = new TreeNode(preorder(index))
            bstFromPreorderWithIndex(index + 1, parent.right)
          } else {
            val newParent = parentList.head
            parentList = parentList.tail
            bstFromPreorderWithIndex(index, newParent)
          }
        }
      }
    }

    bstFromPreorderWithIndex(1, root)
    root
  }
}

//recursive (not tailRec) solution
object Solution2 {
  def bstFromPreorder(preorder: Array[Int]): TreeNode =
    bstFromPreorderWithRange(0, preorder.length - 1, preorder)

  def bstFromPreorderWithRange(start: Int, end: Int, preorder: Array[Int]): TreeNode = {
    if (start > end) null
    else {
      val node = new TreeNode(preorder(start))
      val leftIndexBound = (start to end).find(i => preorder(i) > preorder(start)).getOrElse(end + 1) - 1
      node.left = bstFromPreorderWithRange(start + 1, leftIndexBound, preorder)
      node.right = bstFromPreorderWithRange(leftIndexBound + 1, end, preorder)
      node
    }
  }
}